apps/html-renderer/packages/generic/src/renderHTML.ts

```diff
@@ -0,0 +1,497 @@
+					// Just copy the full video
+					await fs.promises.copyFile(recording.fileName, croppedFilename)
+				} else {
+					// Add margins:
+					boundingBox.x1 -= 10 + (boundingBox.x1 > width * 0.65 ? 10 : 0)
+					boundingBox.x2 += 10 + (boundingBox.x2 < width * 0.65 ? 10 : 0)
+					boundingBox.y1 -= 10 + (boundingBox.y1 > height * 0.65 ? 10 : 0)
+					boundingBox.y2 += 10 + (boundingBox.y2 < height * 0.65 ? 10 : 0)
+
+					logger.verbose(`Saving cropped recording to ${croppedFilename}`)
+					// Generate a cropped video as well:
+					await ffmpeg(logger, [
+						'-hide_banner',
+						'-y',
+						'-i',
+						escapeFilePath(recording.fileName),
+						'-filter:v',
+						`crop=${boundingBox.x2 - boundingBox.x1}:${boundingBox.y2 - boundingBox.y1}:${boundingBox.x1}:${
+							boundingBox.y1
+						}`,
+						escapeFilePath(croppedFilename),
+					])
+					logger.verbose(`Saved cropped recording`)
+				}
+				return croppedFilename
+			},
+		}
+		const cleanupTemporaryFiles = async () => {
+			try {
+				if (recording) {
+					await fs.promises.rm(recording.tmpFolder, { recursive: true })
+				}
+				// Look for old tmp files
+				const oldTmpFiles = await fs.promises.readdir(tempFolder)
+				for (const oldTmpFile of oldTmpFiles) {
+					const oldTmpFileath = path.join(tempFolder, oldTmpFile)
+					const stat = await fs.promises.stat(oldTmpFileath)
+					if (stat.ctimeMs < Date.now() - 60000) {
+						await fs.promises.rm(oldTmpFileath, { recursive: true })
+					}
+				}
+			} catch (e) {
+				// Just log and continue...
+				// eslint-disable-next-line no-console
+				console.error(e)
+			}
+		}
+		if (options.interactive) {
+			await options.interactive(api)
+		} else {
+			await api.waitForLoad()
+
+			const waitForScripts: Promise<void>[] = []
+			const maxDelay = Math.max(...options.scripts.map((s) => s.wait))
+			const afterLoop: (() => void)[] = []
+			for (const script of options.scripts) {
+				await sleep(script.wait)
+
+				if (script.takeScreenshot) {
+					const fileName = await api.takeScreenshot(script.takeScreenshot.name)
+					logger.info(`Screenshot: ${fileName}`)
+				}
+				if (script.fcn) {
+					logger.verbose(`Executing fcn`)
+					await Promise.resolve(script.fcn({ webContents: win.webContents }))
+				}
+				if (script.executeJs) {
+					logger.verbose(`Executing js: ${script.executeJs}`)
+					await api.executeJs(script.executeJs)
+				}
+				if (script.logInfo) {
+					logger.info(script.logInfo)
+				}
+
+				if (script.startRecording) {
+					let videoFilename = 'N/A'
+					const startRecording = async () => {
+						await new Promise<void>((resolve, reject) => {
+							if (!script.startRecording) return
+
+							const startTime = Date.now()
+							const idleFrameTime = Math.max(500, maxDelay)
+							let maxFrameIndex = 0
+							const endRecording = () => {
+								logger.verbose(`Ending recording, got ${maxFrameIndex} frames`)
+								win.webContents.endFrameSubscription()
```

Comment:
I'm a bit worried about how intertwined this code is. The `beginFrameSubscription` for this `endFrameSubscription` happens some 200 lines above? But is also present in `api.stopRecording`? It's really difficult to follow what's controlling what. Maybe one way would be to set up an abstraction to control the frame capturing/recording/ffmpeg functionality and have `renderHTML` control that, instead of talking directly to the BrowserWindow?